src/fs/metadata_log_operations/create_file.hh

```diff
@@ -0,0 +1,127 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2020 ScyllaDB
+ */
+
+#pragma once
+
+#include "fs/metadata_log.hh"
+#include "fs/path.hh"
+
+namespace seastar::fs {
+
+class create_file_operation {
+    metadata_log& _metadata_log;
+    bool _is_directory;
+    sstring _entry_name;
+    file_permissions _perms;
+    inode_t _dir_inode;
+    inode_info::directory* _dir_info;
+    ondisk_create_inode_as_dir_entry_header _ondisk_entry;
+
+    create_file_operation(metadata_log& metadata_log) : _metadata_log(metadata_log) {}
+
+    future<inode_t> perform_1(sstring path, file_permissions perms, bool is_directory) {
+        _is_directory = is_directory;
+        if (is_directory) {
+            while (not path.empty() and path.back() == '/') {
+                path.erase(path.end() - 1, path.end());
+            }
+        }
+
+        _entry_name = last_component(path);
```

Comment:
@psarna Is it OK, to use std::filesystem::path(path).filename().string() instead of last_component(path)?. It would make last_component() redundant, but I don't know the performance implications of using std::filesystem::path...

Reply:
I dug a little through the libstdc++ source code and it seems that std::filesystem::path splits path on every '/' and stores path both as a string and vector of its components that have type std::filesystem::path. So it seems a lot of unnecessary copying.

Reply:
It's still fine for now - we leave microoptimizations for later and focus on core file system implementation. If it ever comes up in perf as a bottleneck, it will be easy to fix.